### PR TITLE
Limit attachment size under 20MB in LiveChatWidget and Telemetry for V2

### PR DESCRIPTION
**Pull request by josuhern**
**Description**
As AMSClient file size restriction 20Mb for images, 300Mb for other files
we need to the sizes of the file,
to prevent failures before to upload the file.

**Solution Proposed**
Adding an extra validation on AttachmentUploadValidatorMiddleware.ts that checks both Dynamics and AMSClient file restrictions and validates the file before submitting it

**Acceptance criteria**
validation for images is up to 20MB
validation for any other attachment 300MB
error should be displayed in the UI

**Test cases and evidence**

1. Image of 10Mb, Files of 50Mb and 100Mb tested
image

2. Failure message after trying to upload a 30Mb image
image

3. Failure after trying to upload 200Mb file
image

Files used
image

### Telemetry section
**Description**
Adding telemetry to attachment Upload Validator Middleware when the validation has an exception

**Solution Proposed**
Adding logs to telemetry for the differnt cases that there could be an exception while validating a file

**Acceptance criteria**
Logs for different case scenarios should be logged in Kusto

**Test cases and evidence**
After trying to upload image over 30Mb, file over 128Mb I got the following logs in telemetry
https://josue.blob.core.windows.net/telemetry/indexv2.html
image
image

**Query**
union lcw_*
| where OrganizationId == '27ff8a63-815a-49f6-b180-b248f806b867'
| where Event == "AttachmentUploadValidatorMiddlewareFailed"

### CODE REVIEW

**Comment by sarojkpr** [comment]
The UI seems little confusing as the 100mb message says "Sent" with an error, so not exactly giving any idea file sent or not sent. Can we use "failed | retry" for this purpose?

File: chat-widget/src/common/Constants.ts [hunk 1]
```diff
@@ -182,6 +182,12 @@ export class WebChatMiddlewareConstants {
     public static readonly adaptiveCard = "AdaptiveCard";
 }
 
+export class AMSConstants {
```
Comment (charliewang95) [inline chat-widget/src/common/Constants.ts:185]: The screenshots seem to be V1's. Please test on V2 and attach evidence

Comment (josuhern) [reply chat-widget/src/common/Constants.ts:185]: yes I used the same screenshots just for making it easy but I have two different testing env 
V1: https://josue.blob.core.windows.net/tasks/indexv1.html
V2: https://josue.blob.core.windows.net/tasks/indexv2.html
 
image

V1 points to drop folder scritps
V2 points to dist folder scritps

I tested in both and I have the same results, that's why I just decided to use same images

File: chat-widget/src/components/webchatcontainerstateful/webchatcontroller/middlewares/storemiddlewares/attachmentUploadValidatorMiddleware.ts [hunk 2]
```diff
@@ -9,23 +9,25 @@ import { NotificationHandler } from "../../notification/NotificationHandler";
...
 * If an attachment is invalid, delete this attachment from the attachments list
 * If the result attachment list is empty, return a dummy action
 */
-const validateAttachment = (action: IWebChatAction, allowedFileExtensions: string, maxUploadFileSize: string, localizedTexts: ILiveChatWidgetLocalizedTexts): IWebChatAction => {
+const validateAttachment = (action: IWebChatAction, allowedFileExtensions: string, maxFileSizeSupportedByDynamics: string, localizedTexts: ILiveChatWidgetLocalizedTexts): IWebChatAction => {
     const attachments = action?.payload?.activity?.attachments;
     const attachmentSizes = action?.payload?.activity?.channelData?.attachmentSizes;
     if (attachments) {
         for (let i = 0; i < attachments.length; i++) {
+            const maxUploadFileSize = getmaxUploadFileSize(maxFileSizeSupportedByDynamics, attachments[i].contentType);
```
Comment (xTEddie) [inline chat-widget/src/components/webchatcontainerstateful/webchatcontroller/middlewares/storemiddlewares/attachmentUploadValidatorMiddleware.ts:25]: nit: `getmaxUploadFileSize` -> `getMaxUploadFileSize`

File: chat-widget/src/components/webchatcontainerstateful/webchatcontroller/middlewares/storemiddlewares/attachmentUploadValidatorMiddleware.ts [hunk 3]
```diff
@@ -58,8 +60,24 @@ const validateFileExtension = (attachment: any, allowedFileExtensions: string):
     return allExtensions.indexOf(fileExtension) > -1;
 };
 
-const validateFileSize = (attachmentSize: string, maxUploadFileSize: string): boolean => {
-    return (maxUploadFileSize && parseInt(maxUploadFileSize) * MBtoBRatio) > parseInt(attachmentSize);
+const validateFileSize = (attachmentSize: string, maxUploadFileSize: number): boolean => {
+    return (maxUploadFileSize * MBtoBRatio) > parseInt(attachmentSize);
+};
+
+const getmaxUploadFileSize = (maxFileSizeSupportedByDynamicsStr:string, contentType: string): number => {
```
Comment (xTEddie) [inline chat-widget/src/components/webchatcontainerstateful/webchatcontroller/middlewares/storemiddlewares/attachmentUploadValidatorMiddleware.ts:67]: nit: Rename to `getMaxUploadFileSize`

Comment (xTEddie) [reply chat-widget/src/components/webchatcontainerstateful/webchatcontroller/middlewares/storemiddlewares/attachmentUploadValidatorMiddleware.ts:67]: nit: spacing

`maxFileSizeSupportedByDynamicsStr:string` -> `maxFileSizeSupportedByDynamicsStr: string`

File: chat-widget/src/components/webchatcontainerstateful/webchatcontroller/middlewares/storemiddlewares/attachmentUploadValidatorMiddleware.ts [hunk 4]
```diff
@@ -58,8 +60,24 @@ const validateFileExtension = (attachment: any, allowedFileExtensions: string):
...
+    const isFileImage = isImage(contentType);
```
Comment (xTEddie) [inline chat-widget/src/components/webchatcontainerstateful/webchatcontroller/middlewares/storemiddlewares/attachmentUploadValidatorMiddleware.ts:68]: We can probably simplify the code to use ternary operator

```js
const amsAttachmentSizeLimit = isImage(contentType)? AMSConstants.maxSupportedImageSize: AMSConstants.maxSupportedFileSize;

// ...

return maxFileSizeSupportedByDynamics < amsAttachmentSizeLimit? maxFileSizeSupportedByDynamics: amsAttachmentSizeLimit;

```

File: chat-widget/src/components/webchatcontainerstateful/webchatcontroller/middlewares/storemiddlewares/attachmentUploadValidatorMiddleware.ts [hunk 5]
```diff
@@ -58,8 +60,24 @@ const validateFileExtension = (attachment: any, allowedFileExtensions: string):
...
+    // Takes the smallest max file size configure betteween AMS and Dynamics Config
+    if (isFileImage){
+        return maxFileSizeSupportedByDynamics > AMSConstants.maxSupportedImageSize ? AMSConstants.maxSupportedImageSize : maxFileSizeSupportedByDynamics;
+    } else {
+        return maxFileSizeSupportedByDynamics > AMSConstants.maxSupportedFileSize ? AMSConstants.maxSupportedFileSize: maxFileSizeSupportedByDynamics;
+    }
+};
+
+const isImage = (contentType: string): boolean => {
+    return AMSConstants.supportedImagesMimeTypes.indexOf(contentType) > -1;
```
Comment (xTEddie) [inline chat-widget/src/components/webchatcontainerstateful/webchatcontroller/middlewares/storemiddlewares/attachmentUploadValidatorMiddleware.ts:80]: `Array.prototype.includes()` can probably be used instead for readability.

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/includes

Comment (josuhern) [reply chat-widget/src/components/webchatcontainerstateful/webchatcontroller/middlewares/storemiddlewares/attachmentUploadValidatorMiddleware.ts]: done

File: chat-widget/src/components/webchatcontainerstateful/webchatcontroller/middlewares/storemiddlewares/attachmentUploadValidatorMiddleware.ts (hunk 4 above)
Comment (josuhern) [reply chat-widget/src/components/webchatcontainerstateful/webchatcontroller/middlewares/storemiddlewares/attachmentUploadValidatorMiddleware.ts]: done

File: chat-widget/src/components/webchatcontainerstateful/webchatcontroller/middlewares/storemiddlewares/attachmentUploadValidatorMiddleware.ts (hunk 2 above)
Comment (josuhern) [reply chat-widget/src/components/webchatcontainerstateful/webchatcontroller/middlewares/storemiddlewares/attachmentUploadValidatorMiddleware.ts]: done

File: chat-widget/src/components/webchatcontainerstateful/webchatcontroller/middlewares/storemiddlewares/attachmentUploadValidatorMiddleware.ts (hunk 3 above)
Comment (josuhern) [reply chat-widget/src/components/webchatcontainerstateful/webchatcontroller/middlewares/storemiddlewares/attachmentUploadValidatorMiddleware.ts]: done

File: chat-widget/src/components/webchatcontainerstateful/webchatcontroller/middlewares/storemiddlewares/attachmentUploadValidatorMiddleware.ts [hunk 6]
```diff
@@ -58,72 +62,119 @@ const validateFileExtension = (attachment: any, allowedFileExtensions: string):
...
+            Event: TelemetryEvent.AttachmentUploadValidatorMiddlewareFailed,
+            Description: "Attachment validation failed",
+            ExceptionDetails: { ErrorDetails: "File provided is empty" }
+        });
         errorMessage = localizedTexts.MIDDLEWARE_BANNER_FILE_IS_EMPTY_ERROR ?? "";
     } else {
+        TelemetryHelper.logActionEvent(LogLevel.ERROR, {
+            Event: TelemetryEvent.AttachmentUploadValidatorMiddlewareFailed,
+            Description: "Attachment validation failed",
+            ExceptionDetails: { ErrorDetails: "Unexpected error: supportedFileExtension=" + supportedFileExtension + " supportedFileSize=" + supportedFileSize + " fileIsEmpty=" + !fileIsEmpty }
```
Comment (charliewang95) [inline chat-widget/src/components/webchatcontainerstateful/webchatcontroller/middlewares/storemiddlewares/attachmentUploadValidatorMiddleware.ts:108]: Please use string interpolation

Comment (charliewang95) [reply chat-widget/src/components/webchatcontainerstateful/webchatcontroller/middlewares/storemiddlewares/attachmentUploadValidatorMiddleware.ts:108]: Here and below

File: chat-widget/src/common/Constants.ts [hunk 7]
```diff
@@ -185,6 +185,12 @@ export class WebChatMiddlewareConstants {
...
+    public static readonly supportedImagesMimeTypes = ["image/jpeg", "image/png", "image/gif", "image/heic", "image/webp"]; //using same MIME types as AMSClient
+    public static readonly maxSupportedImageSize = 20; //AMS max size for images is 20MB
```
Comment (xTEddie) [inline chat-widget/src/common/Constants.ts:190]: nit: Space after `//`

File: chat-widget/src/common/Constants.ts [hunk 8]
```diff
@@ -185,6 +185,12 @@ export class WebChatMiddlewareConstants {
...
+    public static readonly maxSupportedFileSize = 300; //AMS max size limit is 300MB but Dynamics limit is 128MB this gets overwritten on AttachmentUploadValidatorMiddleware.ts line 78
```
Comment (xTEddie) [inline chat-widget/src/common/Constants.ts:191]: Can we not specify the line number? The code base will evolve and the line number won't eventually be true anymore.

Comment (xTEddie) [reply chat-widget/src/common/Constants.ts:191]: `// AMS max file limit outside of supported Images MIME Types.`

**Review by charliewang95 — APPROVED** [review]

**Pull request merged**